blogposts/2020/code-search-turned-code-checker.md

```diff
@@ -0,0 +1,296 @@
+---
+title: Code search turned static code checker
+author: Rijnard van Tonder
+authorUrl: https://twitter.com/rvtond
+publishDate: 2020-09-30T14:00-07:00
+tags: [blog]
+slug: code-search-turned-code-checker
+heroImage: /blog/XXX.png
+published: true
+---
+
+<style>
+  .gatsby-highlight {
+    max-width: 100%;
+    width: 40rem;
+    margin-left: auto;
+    margin-right: auto;
+  }
+  table {
+    width: 40rem;
+    border: none;
+  }
+    table th {
+    border: none;
+  }
+  table td {
+    border: none;
+  }
+  table td:nth-child(2n) {
+      text-align: right;
+  }
+    table th:nth-child(2n) {
+      text-align: right;
+  }
+  table tr:nth-child(2n) {
+    background-color: transparent;
+  }
+</style>
+
+I find static code checkers most valuable when they help teach me better ways to
+code in a language or framework. For example, the Go
+[staticcheck](https://staticcheck.io/docs/checks#SA6005) tool finds expensive
+string comparisons like:
+
+```go
+if strings.ToLower(string1) == strings.ToLower(string2) {
+  ...
+}
+```
+
+and suggests instead:
+
+```go
+if strings.EqualFold(string1, string2) {
+  ...
+}
+```
+
+I find these short-and-sweet replacements are a great way to learn framework
+idioms or library functions, like `strings.EqualFold` in Go.<sup>1</sup> As a
+codebase grows, these small inefficiences, inconsistencies, and missed
+opportunities compound. Code patterns creep in that affect readability and
+performance—[and it matters](https://www.digitalocean.com/blog/how-to-efficiently-compare-strings-in-go/?).
+
+## Productivity workflows: Code checks vs. code search
+
+Code checkers like `staticcheck` typically integrate with continuous integration
+(CI) pipelines, or maybe a pre-commit hook in your development environment.
+Others, like lint checks, typically integrate with editors. These workflows need
+some upfront configuration, the code checks are fixed in place, and productivity
+ensues (🤞).
+
+A code search can also find code snippets to replace with functions like
+`EqualFold`. In practice though, search engines generally treat code as
+plaintext, and can't offer the fidelity of dedicated code checkers. Code
+checkers do more work to gather static information of programs, e.g., parsing
+syntax into trees and using build outputs like types and dependency graphs. It
+takes time to gather this info, it takes knowledge to write checks that use this
+info, and it takes time to hook it into your workflow. These constraints take away the
+flexibility and speed of search workflows when you're simply looking for changes
+to your favorite
+`BananaNutChocolateCake Provider`[↗](https://sourcegraph.com/search?q=BananaNutChocolateCake&patternType=literal).
+
+And yet, I can't shake the idea that there's a middle ground between crude
```

Comment:
To me, this intro leaves out a huge advantage of search over staticcheck, which is scalability. With search+comby, you can execute your find-replace over the entire universe of code, whereas staticcheck requires manual setup/effort for each repository you want to cover. This difference in scale/automaticness means if I'm someone who added a new API, I would totally reach for search+comby to update existing usages of that API to take advantage of my new function. Not as easy with staticcheck, because other repository maintainers have to set it up first. Big difference in friction.

Reply:
I agree. I want this post to be more of a deep-dive and examples of existing checks (I do get into the 'universe of code' bit more clearly in the intro now). The scenario of a custom API is great and I get into that a bit later, but it's difficult tie that benefit into this idea of code checks in the intro.